src/mem/ruby/structures/ALUFreeListArray.cc

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023 The University of Wisconsin
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mem/ruby/structures/ALUFreeListArray.hh"
+
+#include "base/intmath.hh"
+#include "mem/ruby/system/RubySystem.hh"
+#include "sim/cur_tick.hh"
+
+#include <stdio.h>
+
+
+namespace gem5
+{
+
+namespace ruby
+{
+
+/*
+* Behavior:
+*   Requests will go through unless one/both of the following are met:
+*       - There have been more than [numALUs] requests in the current cycle
+*       - The same line has been accessed in the past accessLatency cycles
+*/
+
+ALUFreeListArray::ALUFreeListArray(unsigned int numALUs, Cycles accessLatency,
+                                   RubySystem *rs)
+    : m_ruby_system(rs)
+{
+    this->numALUs = numALUs;
+    this->accessLatency = accessLatency;
+}
+
+bool ALUFreeListArray::tryAccess(Addr addr)
+{
+    uint32_t accesses_this_tick = 0;
+
+    // Remove requests from the tail of the queue that occured more than
+    // accessLatency ticks ago
+    Tick oldestValidRecordStart = curTick() -
+                                ((this->accessLatency) *
+                                m_ruby_system->clockPeriod());
```

Comment:
This is likely to behave in unexpected ways. What happens if the cache (or controller) has a different clock period than the "ruby system"? I.e., you cannot assume all caches are in the same clock domain.

I suggest taking a `Tick` as the `accessLatency` and converting from cycles to ticks in the controller.

Reply:
I think the Tick thing makes sense too.

Reply:
@powerjg Would the clock period not correspond to the ruby_system passed to CacheMemory (which is wrapped as RubyCache) on say, [line 158 of MOESI_AMD_BASE.py](https://github.com/gem5/gem5/blob/develop/configs/ruby/MOESI_AMD_Base.py#L158)? If it doesn't, would BankedArray, which is used by dataArray and tagArray [also seems to be using this ruby_system](https://github.com/gem5/gem5/blob/develop/src/mem/ruby/structures/BankedArray.cc#L98) to get the clock period in ticks also need to be changed?

Reply:
I've updated ALUFreeListArray to take Ticks, rather than Cycles.